_source/_posts/2022-01-13-node-deploy-to-heroku.md

```diff
@@ -0,0 +1,492 @@
+var usersRouter = require("./routes/users");
+
+const session = require("express-session");
+const { ExpressOIDC } = require("@okta/oidc-middleware");
+
+var app = express();
+
+// session support is required to use ExpressOIDC
+app.use(
+  session({
+    secret: "this should be secure",
+    resave: true,
+    saveUninitialized: false,
+  })
+);
+
+const oidc = new ExpressOIDC({
+  issuer: process.env.OKTA_OAUTH2_ISSUER,
+  client_id: process.env.OKTA_OAUTH2_CLIENT_ID,
+  client_secret: process.env.OKTA_OAUTH2_CLIENT_SECRET,
+  appBaseUrl: process.env.APP_BASEURL,
+  redirect_uri: `${process.env.APP_BASEURL}/authorization-code/callback`,
+  scope: "openid profile",
+});
+
+// ExpressOIDC attaches handlers for the /login and /authorization-code/callback routes
+app.use(oidc.router);
+
+// view engine setup
+app.set("views", path.join(__dirname, "views"));
+app.set("view engine", "jade");
+
+app.use(logger("dev"));
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, "public")));
+
+app.use(express.static(__dirname + "/node_modules/bootstrap/dist"));
+
+app.use("/", indexRouter);
+app.use("/users", usersRouter.userRoutes({ oidc: oidc }));
+
+// catch 404 and forward to error handler
+app.use(function (req, res, next) {
+  next(createError(404));
+});
+
+// error handler
+app.use(function (err, req, res, next) {
+  // set locals, only providing error in development
+  res.locals.message = err.message;
+  res.locals.error = req.app.get("env") === "development" ? err : {};
+
+  // render the error page
+  res.status(err.status || 500);
+  res.render("error");
+});
+
+module.exports = app;
+```
+
+Most of this code is boilerplate from the express-generator but the notable addition is the configuration for the Okta oidc.  First, you need to enable `dotenv` at the top of this file.  Next, you configure the `express-session` and add it to your application.  Finally, you can configure your Okta oidc middleware using the variables from `.env`.   
+
+Next, open your `routes/index.js` file and replace the code with the following.
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+/* GET home page. */
+router.get("/", function (req, res, next) {
+  res.render("index", { loggedIn: req.isAuthenticated() });
+});
+
+module.exports = router;
+```
+
+Here are you added a property for the `index.jade` template to use called `loggedIn`.  This property will help the view decide if it should display a `login` or `logout` button.  The rest of the home page will be static so there is no need for a more sophisticated model.
+
+Finally, open your `routes/users.js` file and replace the code there with the code below.
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+function userRoutes(options) {
+  const oidc = options.oidc;
+
+  router.get("/", oidc.ensureAuthenticated(), function (req, res, next) {
```

Comment:
Should this be `/users`?

Reply:
No, the prefix is added in the app.js file.